### PR TITLE
[Feature] Add qwen/glm dspark for mrv1

### PR DESCRIPTION
**Pull request by chenaoxuan**
This PR comes from PR-[11153 ](https://github.com/vllm-project/vllm-ascend/pull/11153), and is based on the latest vllm community for main2main refactoring. The old PR will be closed.
### What this PR does / why we need it?
add Qwen/GLM DSpark speculation method for morel_runner_v1.
**Performance Test**
The first 100 requests from gsm8k dataset, num_speculative_tokens=7, temperature=0 and without topk/topp, output_len=2048

| Model | Spec Method | Acc Length | Acc Rate per Position | Requests per Second | Accuracy |
|-----|-----|-----|-----|-----|-----|
| Qwen3-8B | DFlash | 4.73 | [0.85, 0.71, 0.59, 0.50, 0.42, 0.36, 0.30] | 4.73 | 0.85 |
| Qwen3-8B | **Dspark** | 6.30 | [0.94, 0.87, 0.81, 0.74, 0.69, 0.63, 0.59] | 6.62 | 0.85 |

### Does this PR introduce _any_ user-facing change?
Inherited from dflash_proposer and created dspark_proposer.

### How was this patch tested?
Version Compatibility: Use the vLLM version or commit after July 3.

--speculative-config '{"num_speculative_tokens": K,"method":"dspark","model":"drafter_path"}'
--compilation-config '{"cudagraph_capture_sizes":[bs * K, bs * (K+1)],"cudagraph_mode": "PIECEWISE"}'

Support EAGER or PIECEWISE 

### Next Plan
Through confidenceHead, we can achieve dynamic speculation. However, currently: we have already obtained algorithmic gains, and after applying confidenceHead, the acceptance length has increased by approximately 0.x, but performance has degraded.
We are investigating the reasons (algorithms/framework/operators, etc.).

### Other Things   
- DSpark has two parallel modules: the Qwen/GLM series use DFlash, while the Deepseek series use a modified MTP. This PR only targets the former and serves model_runner_v1.

- vLLM version: v0.23.0
- vLLM main: https://github.com/vllm-project/vllm/commit/1f486d96a17303ce8db8e02be39545b2be338446

### CODE REVIEW

File: vllm_ascend/ops/triton/spec_decode/utils.py [hunk 1]
```diff
@@ -93,6 +93,7 @@ def copy_and_expand_dflash_inputs_kernel_single_grid(
     total_input_tokens,  # tl.int32
     batch_size,  # tl.int32
     HAS_NUM_REJECTED: tl.constexpr = False,
+    IS_DSPARK: tl.constexpr = False,
```
Comment (drslark) [inline vllm_ascend/ops/triton/spec_decode/utils.py]: Please follow vllm's naming, change it to SAMPLE_FROM_ANCHOR.

File: vllm_ascend/spec_decode/llm_base_proposer.py [hunk 2]
```diff
@@ -1218,21 +1220,36 @@ def _run_merged_draft(
...
-                logits, token_indices_to_sample = self._align_tensor_and_indices(
-                    logits,
-                    num_indices,
-                    token_indices_to_sample,
-                    ori_token_indices_to_sample,
-                    is_logits=True,
-                )
-            draft_token_ids = logits.argmax(dim=-1)
+            if self.method == "dspark":
+                raw_logits = self.model.compute_logits(last_hidden_states)
```
Comment (drslark) [inline vllm_ascend/spec_decode/llm_base_proposer.py:1237]: Add more comments here.

File: vllm_ascend/worker/model_runner_v1.py [hunk 3]
```diff
@@ -3929,6 +3936,17 @@ def mock_pass(param1, param2):
             load_model_total_time,
         )
 
+    def _get_dspark_aux_layers_from_config(self) -> tuple[int, ...] | None:
```
Comment (drslark) [inline vllm_ascend/worker/model_runner_v1.py]: Consider merging it with `_get_eagle3_aux_layers_from_config`.

File: vllm_ascend/spec_decode/dspark_proposer.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+            vllm_config,
+            device,
+            runner=runner,
+        )
+
+        blk = 1 + self.num_speculative_tokens
+        self._dspark_seed_buffer = torch.zeros(self.max_batch_size, dtype=torch.int64, device=device)
+        self._dspark_draft_buffer = torch.zeros((self.max_batch_size, blk), dtype=torch.int64, device=device)
+
+    def set_inputs_first_pass(
```
Comment (drslark) [inline vllm_ascend/spec_decode/dspark_proposer.py:31]: Add more comments for `set_inputs_first_pass`.

File: vllm_ascend/spec_decode/dspark_proposer.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        cad.max_seq_len = cad.max_seq_len + num_query_per_req
+        cad.slot_mapping = query_slot_mapping
+        cad.causal = False
+        cad.attn_mask = None
+        cad.attn_state = AscendAttentionState.ChunkedPrefill
+
+        return num_query_total, token_indices_to_sample, cad, None
+
+    @torch.inference_mode()
+    def dummy_run(
```
Comment (drslark) [inline vllm_ascend/spec_decode/dspark_proposer.py:134]: Add more comments for `dummy_run`.

**Review by drslark — CHANGES_REQUESTED** [review]

File: vllm_ascend/ops/triton/spec_decode/utils.py (hunk 1 above)
Comment (chenaoxuan) [reply vllm_ascend/ops/triton/spec_decode/utils.py]: DONE.

File: vllm_ascend/spec_decode/llm_base_proposer.py (hunk 2 above)
Comment (chenaoxuan) [reply vllm_ascend/spec_decode/llm_base_proposer.py:1237]: DONE.

File: vllm_ascend/worker/model_runner_v1.py (hunk 3 above)
Comment (chenaoxuan) [reply vllm_ascend/worker/model_runner_v1.py]: DONE.

**Review by yiz-liu — COMMENTED** [review]
Can we remove this patch?

**Comment by chenaoxuan** [comment]
> Can we remove this patch?

In the init of vllm model_runner, use_eagle() will be enabled, first creating an EagleProposer. This causes _init_parallel_drafting_params to be called prematurely. It cannot be intercepted on the vllm_ascend side.

**Review by yiz-liu — APPROVED** [review]

File: vllm_ascend/spec_decode/dspark_proposer.py (hunk 4 above)
Comment (chenaoxuan) [reply vllm_ascend/spec_decode/dspark_proposer.py:31]: Done.

File: vllm_ascend/spec_decode/dspark_proposer.py (hunk 5 above)
Comment (chenaoxuan) [reply vllm_ascend/spec_decode/dspark_proposer.py:134]: Done.

**Review by drslark — APPROVED** [review]
LGTM. 
I think we can more easily merge the dsv4 dspark into the current dsaprk_proposer.

**Review by MengqingCao — APPROVED** [review]

**Comment by chenaoxuan** [comment]
This PR also received contributions from  @Sawyer117 and @zongzuo: added the cudagraph-safe part (_dspark_seed_buffer / _dspark_draft_buffer, and the seed copy and zero padding logic in set_inputs_first_pass). 
Thanks!
